dynamic_programming/knapsack.c

```diff
@@ -0,0 +1,156 @@
+/**
+ * @file
+ * @brief [01 Knapsack](https://en.wikipedia.org/wiki/Knapsack_problem#0-1_knapsack_problem)
+ * @details
+ * From Wikipedia : The knapsack problem is
+ * Given a set of items, each with a weight and a value, determine
+ * which items to include in the collection
+ * so that the total weight is less than or equal to a given limit and the total value is as large as possible.
+ * @author [aa001R](https://github.com/aa001R)
+ */
+
+#include <assert.h> /// for assert
+#include <stdio.h>
+
+/**
+ * @brief max() macro function
+ */
+#define MAX(a,b) (((a) > (b)) ? (a) : (b))
+
+/**
+ * @brief Maximum capacity of the knapsack
+ */ 
+#define MAXCAPACITY 10
+
+/**
+ * @brief Item structure
+ */
+struct Item {
+	int weight; // weight
+	int value; // value
+};
+
+/**
+ * @brief Print input items function
+ * @param items an array of Item structure
+ * @param cnt the number of items(Item structure)
+ * @returns void
+ */
+void print_input_struct_Items(struct Item items[], int cnt);
+/**
+ * @brief Print the combination of items in the knapsack
+ * @param W the capacity of the knapsack
+ * @param items an array of Item structure
+ * @param cnt the number of items(Item structure)
+ * @param K the 2D array storing the optimal solution
+ * @returns void
+ */
+void print_knapsack_item(int W, struct Item items[], int cnt, int K[][MAXCAPACITY+1]);
+
+/**
+ * @brief 0-1 Knapsack algorithm
+ * @param W the capacity of the knapsack
+ * @param items an array of Item structure
+ * @param cnt the number of items(Item structure)
+ * @returns the maximum value that can be obtained
+ */
+int knapsack(int W, struct Item items[], int cnt) {
+	// Print input items
+	print_input_struct_Items(items, cnt);
+
+	// 2D array to store the optimal solution of subproblems
+	int K[cnt + 1][W + 1];
+
+	// Initialize the first row and column to 0
+	for (int i = 0; i <= cnt; i++)
+		K[i][0] = 0;
+	for (int j = 0; j <= W; j++)
+		K[0][j] = 0;
+
+	// Calculate the optimal solution by selecting items one by one
+	for (int i = 1; i <= cnt; i++) {
+		for (int w = 1; w <= W; w++) {
+			if (items[i - 1].weight <= w) { // If the i-th item can be included in the knapsack
+				int val1 = items[i - 1].value + K[i - 1][w - items[i - 1].weight]; // Value if the i-th item is selected
```

Comment:
nit: can be more intutive, rename the variable to maybe:
`int valueWithPick = ...`
`int valueWithoutPick = ...`

Reply:
Thank you for the feedback! 
I have renamed each variable to make them more intuitive.